gh-92869: ctypes: Add c_time_t

thp opened the pull request:
This implements a simple way of using `ctypes.c_time_t` for 32-bit and 64-bit `time_t` sizes. See #92869 for a discussion on why this is useful and considerations for different platforms. This doesn't take into account platforms where `time_t` is defined as `double` or anything else that's not a signed integer.

<!-- gh-issue-number: gh-92869 -->
* Issue: gh-92869
<!-- /gh-issue-number -->

thp committed:
ctypes: Add c_time_t (Fixes #92869)

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

m-tmatma commented on Lib/ctypes/__init__.py:
```diff
@@ -563,4 +564,9 @@ def DllCanUnloadNow():
+    c_time_t = c_int64
+elif SIZEOF_TIME_T == 4:
+    c_time_t = c_int32
+
```
@thp 
Is it ok not to define else case?

thp replied:
If `time_t` is not 32-bit or 64-bit, `ctypes.c_time_t` will not be defined. I haven't seen a system where it's not 32-bit or 64-bit, but if we encounter such a system, the right type mapping can be added then.

m-tmatma replied:
@thp  
I think you should add to raise an exception on else case to make sure.
It may not happen, but it is a good practice to consider such exceptional case.

thp replied:
PR updated, now raises an exception.

m-tmatma replied:
It seems I had approved this PR yesterday before the PR was updated,
But it was a mistake. I made it on a smart phone.

It seem I can't dismiss  the approval for this repository.

thp committed:
Raise an exception if sizeof(time_t) is not 4 or 8

AA-Turner commented on Doc/library/ctypes.rst:
```diff
@@ -2275,6 +2277,13 @@ These are the fundamental ctypes data types:
+
+   Represents the C :c:type:`time_t` datatype.
+
+   .. versionadded:: 3.X
```
```suggestion
   .. versionadded:: 3.12
```

AA-Turner commented on Lib/ctypes/__init__.py:
```diff
@@ -563,4 +564,11 @@ def DllCanUnloadNow():
+elif SIZEOF_TIME_T == 4:
+    c_time_t = c_int32
+else:
+    raise SystemError("Unexpected sizeof(time_t): %d" % (SIZEOF_TIME_T,))
```
```suggestion
    raise SystemError(f"Unexpected sizeof(time_t): {SIZEOF_TIME_T:d}")
```

AA-Turner commented on Misc/NEWS.d/next/Library/2022-05-17-06-27-39.gh-issue-92869.t8oBkw.rst:
```diff
@@ -0,0 +1 @@
+Added ``c_time_t`` to ``ctypes``, which has the same size as the ``time_t`` type in C (either 32-bit or 64-bit, depending on platform and ABI)
```
```suggestion
Added :class:`~ctypes.c_time_t` to :mod:`ctypes`, which has the same size as
the :c:type:`time_t` type in C (either 32-bit or 64-bit)
```

AA-Turner reviewed (commented):
Seems reasonable, please add tests though.

A

thp committed:
Update Misc/NEWS.d/next/Library/2022-05-17-06-27-39.gh-issue-92869.t8oBkw.rst

Co-authored-by: Adam Turner <9087854+AA-Turner@users.noreply.github.com>

thp committed:
Update Lib/ctypes/__init__.py

Co-authored-by: Adam Turner <9087854+AA-Turner@users.noreply.github.com>

thp committed:
Update Doc/library/ctypes.rst

Co-authored-by: Adam Turner <9087854+AA-Turner@users.noreply.github.com>

thp committed:
ctypes: Add simple c_time_t size testcase

thp commented on Lib/ctypes/test/test_sizes.py:
```diff
@@ -28,6 +28,9 @@ def test_size_t(self):
         self.assertEqual(sizeof(c_void_p), sizeof(c_ssize_t))
 
+    def test_time_t(self):
+        self.assertIn(sizeof(c_time_t), (4, 8))
```
Simple test case added.

As the reason for `c_time_t` is mostly to get the size of the object out of it, `test_sizes.py` is probably the best place to test for this. This test is mostly about checking for the existence of the field, and that it's 4 or 8. If/when platforms are found that have a differently-sized `c_time_t`, this test can be updated.

In the future, if we find a platform where `c_time_t` is of a different type (different size or even non-integer type), some other tests could be added.

m-tmatma replied:
Why don't you compare `sizeof(c_time_t)` with `SIZEOF_TIME_T`?

AA-Turner replied:
It seems like you should be able to use an independent version from `_testcapi`.

A

thp replied:
(closing this conversation as resolved using `SIZEOF_TIME_T`, did not consider `_testcapi` at the moment)

thp committed:
Fix ctypes documentation with C time() function return type

thp committed:
Fix typo

thp committed:
Use SIZEOF_TIME_T for c_time_t test case

thp committed:
Docs: Use rand() instead of time() for function call example

thp committed:
Reword documentation after moving sections

thp committed:
Merge remote-tracking branch 'origin/main' into ctypes_c_time_t

AA-Turner reviewed (approved):
This looks good from my perspective, but I'm not a ctypes specialist. I've asked if a core dev can review.

A

gpshead committed:
use = instead of :d as the error message format.

kumaraditya303 reviewed (approved):
LGTM

gpshead committed:
End sentence with a .

I also removed the 32 or 64 bit wording as that isn't really important. it abstracts that from the user to use the right size.
